src/ethernet/RsRtspClient.cpp

```diff
@@ -0,0 +1,613 @@
+{
+  unsigned res = this->sendGetParameterCommand(*this->m_scs.m_session, this->continueAfterGETCOMMAND, (const char *)&t_opt);
+  // wait for continueAfterPLAY to finish
+  std::unique_lock<std::mutex> lck(m_commandMtx);
+  m_cv.wait_for(lck, std::chrono::seconds(RTSP_CLIENT_COMMANDS_TIMEOUT_SEC), [this] { return m_commandDone; });
+  // for the next command
+  if(!m_commandDone)
+  {
+    RsRtspReturnValue err = 
+      {RsRtspReturnCode::ERROR_TIME_OUT,"client time out"};
+    throw std::runtime_error(format_error_msg(__FUNCTION__, err));
+  }
+  m_commandDone = false;
+  if(m_lastReturnValue.exit_code!=RsRtspReturnCode::OK)
+  {
+    throw std::runtime_error(format_error_msg(__FUNCTION__,m_lastReturnValue));
+  }
+  return m_lastReturnValue.exit_code;
+}
+
+void schedulerThread(RsRTSPClient *t_rtspClientInstance)
+{
+  std::unique_lock<std::mutex> lk(t_rtspClientInstance->getTaskSchedulerMutex());
+  t_rtspClientInstance->envir().taskScheduler().doEventLoop(&t_rtspClientInstance->getEventLoopWatchVariable());
+  lk.unlock();
+}
+
+void RsRTSPClient::initFunc(MemoryPool *t_pool)
+{
+  std::thread thread_scheduler(schedulerThread, this);
+  thread_scheduler.detach();
+  m_memPool = t_pool;
+}
+
+void RsRTSPClient::setDeviceData(DeviceData t_data)
+{
+  m_deviceData = t_data;
+}
+std::vector<IpDeviceControlData> controls;
+
+std::vector<IpDeviceControlData> RsRTSPClient::getControls()
+{
+  this->sendOptionsCommand(this->continueAfterOPTIONS);
+
+  // wait for continueAfterOPTIONS to finish
+  std::unique_lock<std::mutex> lck(m_commandMtx);
+  m_cv.wait_for(lck, std::chrono::seconds(RTSP_CLIENT_COMMANDS_TIMEOUT_SEC), [this] { return m_commandDone; });
+  // for the next command
+  if(!m_commandDone)
+  {
+    RsRtspReturnValue err = 
+      {RsRtspReturnCode::ERROR_TIME_OUT,"client time out"};
+    throw std::runtime_error(format_error_msg(__FUNCTION__, err));
+  }
+  m_commandDone = false;
+  
+  if(m_lastReturnValue.exit_code!=RsRtspReturnCode::OK)
+  {
+    throw std::runtime_error(format_error_msg(__FUNCTION__,m_lastReturnValue));
+  }
+  //TODO: return the controls at argument
+  return controls;
+}
+
+/*********************************
+ *          CALLBACKS            *
+ *********************************/
+
+// TODO: Error handling
+void RsRTSPClient::continueAfterDESCRIBE(RTSPClient *rtspClient, int resultCode, char *resultString)
+{
+  UsageEnvironment &env = rtspClient->envir();                  // alias
+  StreamClientState &scs = ((RsRTSPClient *)rtspClient)->m_scs; // alias
+  RsRTSPClient *rsRtspClient = ((RsRTSPClient *)rtspClient);    // alias
+  
+  if (NULL!=resultString)
+    rsRtspClient->m_lastReturnValue.msg = resultString;
+  rsRtspClient->m_lastReturnValue.exit_code = (RsRtspReturnCode)resultCode;
+
+  //TODO: take logic out of the callback? 
+  do
+  {
+    if (resultCode != 0)
+    {
+      env << "Failed to get a SDP description: " << resultString << "\n";
+      delete[] resultString;
```

Comment:
Deleting memory allocated elsewhere should be reserved for extreme cases. Additionally the same member is used above in `rsRtspClient->m_lastReturnValue.msg = resultString;` which may lead to dangling ptr.

Reply:
we do it due to Live555 implementation, in order to fix it we will need to do risky changes in the Live555.
We wont have dangling ptr because  m_lastReturnValue.msg is string and not char*